src/lsp-server.ts

```diff
@@ -38,6 +38,52 @@ import { computeCallers, computeCallees } from './calls';
 import { IServerOptions } from './utils/configuration';
 import { TypeScriptVersion, TypeScriptVersionProvider } from './utils/versionProvider';
 import { TypeScriptAutoFixProvider } from './features/fix-all';
+import { LspClient, ProgressReporter } from './lsp-client';
+
+class ServerInitializingIndicator {
+    private _loadingProject?: { project: string | undefined; resolve: () => void; reject: () => void; };
+    private _progressReporter?: ProgressReporter;
+
+    constructor(private lspClient: LspClient) {}
+
+    public reset(): void {
+        if (this._loadingProject) {
+            this._loadingProject.reject();
+            this._loadingProject = undefined;
+            if (this._progressReporter) {
+                this._progressReporter.end();
+                this._progressReporter = undefined;
+            }
+        }
+    }
+
+    /**
+     * Signal that a project has started loading.
+     */
+    public startedLoadingProject(projectName: string | undefined): void {
+        // TS projects are loaded sequentially. Cancel existing task because it should always be resolved before
+        // the incoming project loading task is.
+        this.reset();
+
+        this._progressReporter = this.lspClient.createProgressReporter();
+        this._progressReporter.begin('Initializing JS/TS language features');
+
+        new Promise<void>((resolve, reject) => {
+            this._loadingProject = { project: projectName, resolve, reject };
+        });
+    }
+
+    public finishedLoadingProject(projectName: string | undefined): void {
+        if (this._loadingProject && this._loadingProject.project === projectName) {
```

Comment:
Hah, I wanted to suggest that you could use the `this._loadingProject?.projec` instead of `(this._loadingProject && this._loadingProject.project === projectName`,
but apparently ts will still complain bellow.

![image](https://user-images.githubusercontent.com/22029477/145105921-36cf19c2-f0cd-4005-8725-4ec18e203d51.png)

Reply:
just simplified it since creating a promise was pointless in our case

Reply:
but yes, it wouldn't be the same logic when using optional chaining since it could then evaluate to true when `undefined === undefined`